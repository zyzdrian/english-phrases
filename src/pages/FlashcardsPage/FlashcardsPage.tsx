import React from 'react';
import { RouteComponentProps } from '@reach/router';
import styled from 'styled-components';
import { CSSTransition, SwitchTransition } from 'react-transition-group';
import { BottomMenu, Phrase, TopBar } from '../../components';
import useDailyPhrases from '../../hooks/useDailyPhrases';

const StyledFlashcardsPage = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  min-height: 100vh;
  overflow: hidden;
`;

const transitionName = 'slide';

const Container = styled.section`
position: relative;
padding: 0 24px;
max-width: 600px;
width: 100%;

&.${transitionName}-appear {
  left: -100%;
}

&.${transitionName}-appear-active {
  left: 0;
  transition: left 300ms ease-in;
}
&.${transitionName}-enter {
  left: -100%;
}

&.${transitionName}-enter-active {
  left: 0;
  transition: left 300ms ease-in;
}
&.${transitionName}-exit {
  left: 0;
}

&.${transitionName}-exit-active {
  left: 100%;
  transition: left 300ms ease-out;
}
`;

interface IFlashcardsPageProps extends RouteComponentProps {}

export const FlashcardsPage: React.FC<IFlashcardsPageProps> = () => {
    const { dailyPhrases, currentPhrase, currentPhraseIndex, isCurrentPhraseFlipped, actions } = useDailyPhrases();

    return (
        <StyledFlashcardsPage>
            <TopBar title={`Progress: ${currentPhraseIndex + 1}/${dailyPhrases.length}`} />
            {!!dailyPhrases.length && (
                <SwitchTransition>
                    <CSSTransition timeout={300} classNames={transitionName} unmountOnExit key={currentPhrase.id}>
                        <Container>
                            <Phrase phrase={currentPhrase} onShow={actions.enableMarking} isFlipped={isCurrentPhraseFlipped} />
                        </Container>
                    </CSSTransition>
                </SwitchTransition>
            )}
            <BottomMenu
                markAsKnown={actions.markAsKnown}
                markAsNotKnown={actions.markAsNotKnown}
                rollbackState={actions.undo}
                isRollbackDisabled={!actions.canUndo}
                areButtonsDisabled={actions.canMark}
                enableButtons={actions.enableMarking}
            />
        </StyledFlashcardsPage>
    );
};
