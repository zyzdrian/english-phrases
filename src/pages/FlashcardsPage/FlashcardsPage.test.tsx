import React from 'react';
import { render } from '@testing-library/react';
import { FlashcardsPage } from './FlashcardsPage';
import '@testing-library/jest-dom/extend-expect';

jest.mock('../../hooks/useDailyPhrases', () => () => ({
    dailyPhrases: [
        {
            id: '1',
            name: 'Test phrase',
            appropriate: 'Test appropriate',
            chapterId: 1,
            meanings: ['Meaning'],
            bucketIndex: 1
        },
        {
            id: '2',
            name: 'Test phrase 2',
            appropriate: 'Test appropriate',
            chapterId: 1,
            meanings: ['Meaning 1', 'Meaning 2'],
            bucketIndex: 1
        }
    ],
    setDailyPhrases: jest.fn(),
    currentPhrase: {
        id: '1',
        name: 'Test phrase',
        appropriate: 'Test appropriate',
        chapterId: 1,
        meanings: ['Meaning'],
        bucketIndex: 1
    },
    currentPhraseIndex: 0,
    isCurrentPhraseFlipped: false,
    actions: {
        markAsKnown: jest.fn(),
        markAsNotKnown: jest.fn(),
        undo: jest.fn(),
        canUndo: false,
        canMark: false,
        enableMarking: jest.fn()
    }
}));

describe('<FlashcardsPage />', () => {
    it('should return as a default phrases page', () => {
        const { container } = render(<FlashcardsPage />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
