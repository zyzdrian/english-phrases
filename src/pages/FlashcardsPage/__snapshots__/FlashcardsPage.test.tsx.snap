// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlashcardsPage /> should return as a default phrases page 1`] = `
<div
  class="sc-kAzzGY cYQrAh"
>
  <div
    class="sc-VigVT cybuwg"
  >
    <img
      alt="menu"
      class="sc-jTzLTM kMqZix"
      data-testid="hamburger"
      src="menu.svg"
    />
    <div
      class="sc-fjdhpX jmFvEE"
      data-testid="title"
    >
      Progress: 1/2
    </div>
    <span
      style="width: 24px;"
    />
  </div>
  <section
    class="sc-chPdSV cfASoy"
  >
    <div
      class="sc-bxivhb eRFxNn"
    >
      <div
        class="sc-ifAKCX jJulbc"
      >
        <div
          class="sc-EHOje kGzamQ"
          data-testid="front"
        >
          <div
            class="sc-htoDjs sc-dnqmqq bhPwuV"
          >
            <div
              class="sc-iwsKbI dKyiiK"
              data-testid="name"
            >
              Test phrase
            </div>
          </div>
        </div>
        <div
          class="sc-bZQynM eprqWe"
          data-testid="back"
        >
          <div
            class="sc-htoDjs sc-gZMcBi bSzNzK"
          >
            <div
              class="sc-gqjmRU eJtqmJ"
              data-testid="meaning-0"
            >
              Meaning
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="sc-bdVaJa kafHxm"
    data-testid="bottom-menu"
  >
    <button
      class="sc-bwzfXH gtkPbW"
      data-testid="i-dont-know-button"
    >
      I don't know
    </button>
    <button
      class="sc-bwzfXH gtkPbW"
      data-testid="i-know-button"
    >
      I know
    </button>
  </div>
</div>
`;
